Only enable regexp and literal matching styles by default

minad opened the pull request:
* Avoid performance issues with initialism when filtering long candidates (#39)
* README: Update default matching styles
* README: Document `orderless-define-completion-style`
* README: Update `without-if-bang` such that it ignores "!"

Good idea or not? I am not sure if you are okay with adjusting the defaults. In any case it would be nice to document `orderless-define-completion-style` in the readme.

oantolin commented:
I'm happy to change the defaults. I've never liked or used them in my personal configuration. I made those the defaults to match  prescient (do they match prescient?), or maybe because someone suggested them.

I do worry about the odd change to `without-if-bang` which gives it a seemingly odd, inconsistent behavior. Is that a good idea? If you are doing it to avoid suddenly having no matches when you first type `!`, I think the alternative of suddenly matching a literal `!` and then doing something completely different at the next keystroke is just as jarring.

Also, did you decide to indent the body of `flex-if-twiddle` 3 spaces rather than 2?

minad commented:
> I'm happy to change the defaults. I've never liked or used them in my personal configuration. I made those the defaults to match prescient (do they match prescient?), or maybe because someone suggested them. I can't really remember now.

Okay. I think it is better to go with literal+regexp as defaults then since then we are less likely to observe performance problems. I am using `orderless-define-completion-style` in my config to enable initialisms for symbols and a few other categories, where initialisms work well. I only found #50 because of that.

> I do worry about the change to without-if-bang which gives it a seemingly odd, inconsistent behavior. Is that a good idea? If you are doing it to avoid suddenly having no matches when you first type !, I think the alternative of suddenly matching a literal ! and then doing something completely different at the next keystroke is just as jarring.

With Selectrum and Vertico it is a good idea to ignore single "!". Not matching the empty string by itself is not useful. If you type "!", Vertico and Selectrum will show zero candidates until you type the next character, so it is better to ignore the single bang. But if you don't want to have this here, just remove it from the PR. It is okay if we don't add special tricks here which are mainly targeted at some special UIs.

> Also, did you decide to indent the body of flex-if-twiddle 3 spaces rather than 2?

Fixed it.

oantolin commented:
I'm sorry. I did't read the code carefully enough. I thought you were matching a single `!` literally, but instead you match an empty string. I do agree with that change. (Although I seem to remember you can now return nil to ignore.)

minad commented:
> Although` I seem to remember you can now return nil to ignore.

Could be. We changed this when we added the check for invalid regexps? I think this empty bang matcher from my config predates that fix.

oantolin commented:
What do you think about making the `without-if-bang` dispatcher part of the default configuration?

minad commented:
> What do you think about making the without-if-bang dispatcher part of the default configuration?

Right now there are no default style dispatchers are they? For this reason I am using this ugly monolithic dispatcher:

~~~ elisp
(defun dm/orderless-dispatch (pattern _index _total)
  (cond
   ;;((and completion-in-region-mode (= index 0)) `(orderless-regexp . ,(concat "^" pattern)))
   ;; Ensure that $ works with Consult commands, which add disambiguation suffixes
   ((string-suffix-p "$" pattern) `(orderless-regexp . ,(concat (substring pattern 0 -1) "[\x100000-\x10FFFD]*$")))
   ;; File extensions
   ((string-match-p "\\`\\.." pattern) `(orderless-regexp . ,(concat "\\." (substring pattern 1) "[\x100000-\x10FFFD]*$")))
   ;; Ignore single !
   ((string= "!" pattern) `(orderless-literal . ""))
   ;; Without literal
   ((string-prefix-p "!" pattern) `(orderless-without-literal . ,(substring pattern 1)))
   ((string-suffix-p "!" pattern) `(orderless-without-literal . ,(substring pattern 0 -1)))
   ;; Initialism matching
   ((string-prefix-p "`" pattern) `(orderless-initialism . ,(substring pattern 1)))
   ((string-suffix-p "`" pattern) `(orderless-initialism . ,(substring pattern 0 -1)))
   ;; Literal matching
   ((string-prefix-p "=" pattern) `(orderless-literal . ,(substring pattern 1)))
   ((string-suffix-p "=" pattern) `(orderless-literal . ,(substring pattern 0 -1)))
   ;; Flex matching
   ((string-prefix-p "~" pattern) `(orderless-flex . ,(substring pattern 1)))
   ((string-suffix-p "~" pattern) `(orderless-flex . ,(substring pattern 0 -1)))))
~~~

Maybe it would work if orderless supplied a few of those by default if we find a consistent naming scheme. orderless-style-flex-twiddle, orderless-style-literal-equal, orderless-style-initialism-apostrophe, orderless-style-without-literal-bang. But given the variety of dispatchers I think it is perfectly okay to not provide any dispatchers by default. I think it is even better to provide none instead of picking out a few special ones. As soon as we start to add sth like orderless-style-flex, we might want to add a configuration variable, which allows configuring the flex character. I think I am more happy with keeping Orderless barebone. It is good as is!

minad commented:
> but feel that maybe the orderless-without-literal matching style is a special case in that it really isn't useful without a dispatcher.

That's true. But I think this justification is pretty weak. From what I've seen, ever orderless user creates their own style dispatcher. I would not muddy these waters by providing some preconfigured dispatchers.

minad commented:
And well obviously - as soon as you create your own dispatchers, adding the without bang dispatcher is not costly.

oantolin commented:
Yeah, OK, that sounds reasonable. Though if we expect people to create dispatchers I'm not sure why we should care what the default matching styles are. Wouldn't it be reasonable to expect that people will configure those too?

minad commented:
> Though if we expect people to create dispatchers I'm not sure why we should care what the default matching styles are. Wouldn't it be reasonable to expect that people will configure those too?

Actually, I had thought about proposing to remove the default orderless completion style and the configuration variables altogether. Then one would rewrite `orderless-define-completion-style` as a `cl-defmacro` taking some keyword arguments `:matching-styles`, `:style-dispatcher`, ... The user would then have to define their own style before being able to use orderless. But maybe that wouldn't be very user friendly or good elisp-style, in contrast to the current design with the dynamically bound configuration variables :)

oantolin commented:
Oh, that change is a little too radical for me, and a little too Javaesque. I'd have to stop saying that orderless is a completion style and start saying it is a `CompletionStyleFactory`.

minad commented:
Indeed `orderless-define-completion-style` should be rewritten to return a completion style and be renamed to `orderless-completion-style-factory`. But a factory is just a function :-P
